### PR TITLE
Relation dynamic finders

### PR DESCRIPTION
**Pull request by ipundit**
Implemented dynamic finders for Relation:

```php
Author::select('author_id')->find_by_name('Tito');
```

This greatly reduces the need for where clauses, while avoiding the performance hit of retrieving all the columns for a row as in:

```php
Author::find_by_name('Tito');
```

### CODE REVIEW

**Comment by shmax** [comment]
I'll have to dig into this more deeply later, but are you sure we should support this? Modern versions of Rails seem to encourage the use of `where` over dynamic finders and I can't seem to find any samples of this kind of thing in the modern documentation. Were you able to find anything?

**Comment by ipundit** [comment]
No, but this is quite a useful feature that I would use everywhere in my production code when moving to version 2. The new `find()` will return an array or just a model or null depending on the input, thus a lot of my code thus breaks in version 2, but you made the choice to do that and the many users of version 1 will have to update their code to match.

If you always have to check for an array or a single model throughout your code, it gets tedious and error prone. The workaround I came up with, and I would suggest putting in the documentation is to move to `find_by_syntax()` which will always return just a Model or null.  But, the downside of moving to dynamic finders is the loss of `select()` for performance purposes, which is now fixed.

Also, dynamic finders prevent sql injection attacks whereas where() does not in all cases. I don't see any indication in the documentation where dynamic finders are being deprecated, nor would I expect there to be as it's a readable convenience for Ruby on Rails and ActiveRecord.

**Comment by shmax** [comment]
> No, but this is quite a useful feature that I would use everywhere in my production code when moving to version 2. The new `find()` will return an array or just a model or null depending on the input, thus a lot of my code thus breaks in version 2, but you made the choice to do that and the many users of version 1 will have to update their code to match.


Wait, hang on--the new `find` works pretty much like the old one, doesn't it? It returns `Model|array<Model>`. It only returned `null` if you passed in 'first' or 'null', so yes, you would have to refactor those usages. Is that what you meant?
 
> If you always have to check for an array or a single model throughout your code, it gets tedious and error prone. 

That's not the case, and never was. You don't have to check if the return type is single or array, you have prior knowledge of what is going to come out the other end based on what you pass in:

```php
// always single model (or exception):
$book = Book::find(1);

// always array (or exception)
$books = Book::find([1,2,3]):

// always single model or null
$book = Book::first()
```

> The workaround I came up with, and I would suggest putting in the documentation is to move to `find_by_syntax()` which will always return just a Model or null. But, the downside of moving to dynamic finders is the loss of `select()` for performance purposes, which is now fixed.

I'm not sure what you mean. As far as I can tell, all of the `find_by_` methods should return a single record, or null.


> I don't see any indication in the documentation where dynamic finders are being deprecated, nor would I expect there to be as it's a readable convenience for Ruby on Rails and ActiveRecord.

Okay, then sounds like we can go with it.

**Comment by shmax** [comment]
Please make sure to implement the proper static checking extensions. You can basically clone `ModelStaticMethodReflection` (should be able to DRY in a base class or trait). Once you have that, please add a new file (eg. "RelationDynamicFindBy") to test/phpstan, and make sure that `composer stan` passes.

File: lib/Model.php [hunk 1]
```diff
@@ -688,16 +688,15 @@ public function &read_attribute(string $name)
...
-        if ($relationship = $table->get_relationship($name)) {
+        $relationship = $table->get_relationship($name);
+        if (null !== $relationship) {
             $this->__relationships[$name] = $relationship->load($this);
-
-            return $this->__relationships[$name];
         }
 
-        return null;
+        return array_key_exists($name, $this->__relationships) ? $this->__relationships[$name] : null;
```
Comment (shmax) [inline lib/Model.php:699]:
```suggestion
        return $this->__relationships[$name] ?? null;
```

File: lib/Relation.php [hunk 2]
```diff
@@ -106,6 +107,89 @@ public function none(): Relation
...
+     * ```
+     *
+     * @param $method Name of method
+     * @param $args   Method args
+     *
+     * @throws ActiveRecordException
+     *
+     * @see find
+     */
+    public function __call(string $method, mixed $args): mixed
```
Comment (shmax) [inline lib/Relation.php:146]: If you're going to move the docs here, then please remove them from `Model` so they're not repeated.

Comment (ipundit) [reply lib/Relation.php:146]: Yeah, documentation needs work. We're inconsistent all over the place. I think once a documentation generator is chosen and put in place, then we can clean up things, remove duplication, etc. In this particular case for example, should Model refer to Relation's function header or vice versa. And why? What's the convention or reasoning?

Comment (shmax) [reply lib/Relation.php:146]: Well, the only problem I see is that all of the relation-forwarded calls on `Model` are static, whereas the methods on `Relation` aren't, so the samples we show are a bit weird. But, if we have that problem then so does Ruby on Rails. They seem to gloss over it, so I suppose we can, as well, but I suppose it wouldn't hurt to make sure that our samples show a sprinkling of both usages (static and non-static).

Comment (shmax) [reply lib/Relation.php:146]: So, I guess for consistencie's sake, you can go ahead and remove that block on `Model`, like we do for everything else that delegates to `Relation`.

Comment (ipundit) [reply lib/Relation.php]: Done

File: lib/Model.php (hunk 1 above)
Comment (ipundit) [reply lib/Model.php]: Done

**Comment by ipundit** [comment]
> Please make sure to implement the proper static checking extensions. You can basically clone `ModelStaticMethodReflection` (should be able to DRY using a trait like I did for `ModelDynamicStaticMethodReturnTypeReflection` and `RelationDynamicMethodReturnTypeReflection` ). You'll need to add a few lines to phpstan.neon.dist. Once you have that, please add a new file (eg. "RelationDynamicFindBy" modified from "ModelDynamicFindBy") to test/phpstan, and make sure that `composer stan` passes.
I do not know how to do this, nor do I even know what is the expected outcome of implementing the above. How does that change the code in Relation.php? I think the easiest way to resolve this is for you to do it this time and show me the diffs so I know what to do next time.

**Comment by shmax** [comment]
Did you mean to close this?

**Comment by shmax** [comment]
> I do not know how to do this, nor do I even know what is the expected outcome of implementing the above. 

It's all about the static checking. We already had this set up for Model, such that PHPStan would now to flag this kind of thing as an error:

```php
$book = Book::find_by_name("Walden");
if(is_array($book)) {  // static analysis error! $book can't be an array, so there is no point in testing it as one
```

This kind of magic is currently achieved through the PHPStan reflection class I authored a few weeks back, `ModelStaticMethodReflection`.

Now that you've migrated to `Relation`, we have to make sure PHPStan knows about it. Since the only difference between the find_by_ calls on `Relation` and the ones currently on `Model` is that one is static and one is not, you're within copy-pasting distance of the solution (the trait DRY thing is a nice bonus, but it can come later if you get lost).

> How does that change the code in Relation.php?

It wouldn't. All this work is done in PHPStan extensions. See the files in `lib/PhpStan` to get an idea. And note that these files are activated by pairing their names with their tags in phpstan.neon.dist.

> I think the easiest way to resolve this is for you to do it this time and show me the diffs so I know what to do next time.

Easiest for you, maybe. 😄 But no worries, I'll take care of it this time.

File: lib/Relation.php [hunk 3]
```diff
@@ -721,39 +811,51 @@ public function reverse_order(): array
...
+    }
+
+    /**
+     * Converts relation objects to array with a given options.
+     *
+     * @param RelationOptions $options
+     *
+     * @return array<TModel> All the rows that matches query. If no rows match, returns []
+     */
+    private function to_a_withOptions(array $options): array
```
Comment (shmax) [inline lib/Relation.php:850]: This is a bit awkward, isn't it? What are you fixing? If it's related to polluting the user's options with "order" in the case of "having", then you probably won't be surprised to hear that my advice is going to be to move that logic to `Table::options_to_sql`, and keep the rest of this clean.

Comment (ipundit) [reply lib/Relation.php:850]: It's to:

1) Deny the user to be able to call `Author::to_a($myHandCraftedOptionsThatWouldHaveToBeInputChecked)`
2) Fix the bugs in find_by(), first(), last(), and take(). There are new tests in the test classes which trigger the bug of clobbering $options, but this submission fixes those bugs.

Comment (shmax) [reply lib/Relation.php:850]: > 1. Deny the user to be able to call `Author::to_a($myHandCraftedOptionsThatWouldHaveToBeInputChecked)`

Is that a thing? There's no static `to_a` on `Model` and the `to_a` on `Relation` doesn't take any arguments. But let me take a look at your tests...

Comment (shmax) [reply lib/Relation.php:850]: Ah, yep, those tests show the problem, for sure...

Comment (shmax) [reply lib/Relation.php:850]: Maybe a bit simpler?

https://github.com/php-activerecord/activerecord/compare/how-about-this?expand=1

**Comment by ipundit** [comment]
```php
    public function testDoesNotClobberOrderOrLimit()
    {
        $rel = Author::limit(2)->order('author_id DESC');
//        $this->assertEquals(1, $rel->first()->author_id);
```
The line should not be commented and the test must pass. `first()` is defined to return the first record in the table, despite `order('author_id DESC')`. Or should the spec be changed to?:
```php
$author = Author::order('author_id DESC')->first();
$author2 = Author::order('author_id ASC')->last();

$this->assertEquals($author, $author2);
```

As for your other changes, ```clone $this``` would be less performant than the current implementation of cloning just ```$this->options``` rather than the entire ```$this``` object. Also, what does it mean to clone ```$this->generator``` and ```$this->table```? If those classes have a database connection somewhere, or some kind of other resource where a shallow clone would be problematic, then this could lead to difficult to find bugs. I would rather only clone ```$this->options``` if I'm only changing ```$this->options``` and nothing else.

**Comment by shmax** [comment]
> The line should not be commented and the test must pass. first() is defined to return the first record in the table, despite

Yes, yes, the tests work even with that line uncommented.

> Also, what does it mean to clone $this->generator and $this->table

If the tests pass, then the tests pass. 😄 

But if you don't like it, we can go with your code and I'll sort it out later.

**Review by shmax — APPROVED** [review]

**Comment by shmax** [comment]
Thanks!

**Pull request merged**